### DISCUSSION TITLE
One requests response type

### DISCUSSION THREAD

**Discussion by kolserdav** [atherdon/dash-back#15] [Show and tell]
Проблему описывал [тут](https://github.com/atherdon/dash-back/blob/main/docs/CHANGELOG.md).
Суть в том, что была задача сделать удобный API для фронтенда, мной было принято решение, в случае ошибки на сервере возвращать человеко читаемое, но в то же время не просто 'Some went wrong' сообщение клиенту, а также добавить третий тип статуса ответа `warning` если ошибки нет, но требуется действие (неверный запрос или пустой ответ). При работе с базой, легко обработать возможные исключения и вернуть то что нужно. Но параметры запроса на GraphQL можно отследить только на посреднике. Однако библиотек для этого ещё нет. Модуль, который я запилил, стал слишком сложным при добавлении вложенных типов, был удален в версии 2.0.0 и его допиливание сейчас не рассматривается (там только если либу делать).
Но требуется решение, пока не начат фронт и пока не много моделей на беке:
Вопрос такой, раз не получилось сделать все овтеты едиными, значит может и ошибки базы тоже `Some went wrong` возвращать. Так вижу: пользователю не сильно интересно, что именно пошло не так, зато на фронте будет одна проверка успешности ответа вместо двух.

Однако если бы всё было просто, я бы просто убрал эти "костыли" и пусть будет "что-то пошло не так", НО `prisma` имеет недостаток при обновлении модели базы, если передан `id` несуществующего объекта призма выбрасывает исключение, и если попытка изменить несуществующий объект то тоже ошибка.

Тут меня осенило, всё правильно, пошел ка я удалять свои костыли)

**Comment by kolserdav** [comment]
После удаления лишнего, обработка ошибок `prisma` стала бессмысленной, тоже удалю.

Оказывается использование `graphql` избавляет от таких пережитков прошлого как ручной отлов ошибок на сервере.

**Reply by atherdon** [reply]
еще бы js программисты не накручивали на чистый graphql свои плагины - которые немного убивают логику создания агностик-языка для баз данных :) но тогда они не будут js программистами

**Reply by kolserdav** [reply]
Точнее не просто graphql а `apollo-server-express`

**Reply by atherdon** [reply]
не открывал его, на там наверное вообще бида-бида

**Reply by kolserdav** [reply]
> еще бы js программисты не накручивали на чистый graphql свои плагины - которые немного убивают логику создания агностик-языка для баз данных :) но тогда они не будут js программистами

Это точно, сейчас думаю как бы типы свести у них, чтобы можно было делать не:
```typescript
return {
      id: result.id,
      name: result.name,
      edited: result.edited || 0,
      editedPercent: result.editedPercent || 0,
      published: result.published || 0,
      publishedPercent: result.publishedPercent || 0,
      rejected: result.rejectedPercent || 0,
      rejectedPercent: result.rejectedPercent || 0,
      created: result.created?.toISOString() || '',
      updated: result.updated?.toISOString() || '',
    };
```
А просто:
```typescript
return result;
```
Чтобы данные prisma типов типы автоматически преобразовывались в graphq типы, вроде где-то видела такое, сразу с наскоку не получилось въехать)

**Reply by kolserdav** [reply]
видел*

**Reply by kolserdav** [reply]
> не открывал его, на там наверное вообще бида-бида

Я про `apollo-server` на собеседованиях узнал, в компаниях его почему-то юзают. Не вникал, но думаю неспроста)

**Reply by kolserdav** [reply]
> Это точно, сейчас думаю как бы типы свести у них, чтобы можно было делать не:

у prisma и graphql

**Reply by atherdon** [reply]
> > Это точно, сейчас думаю как бы типы свести у них, чтобы можно было делать не:
> 
> у prisma и graphql

так может в репозитории призмы написать им такое предложение - чтобы они это реализовали в следующих релизах. а вам будет плюсик в карму :)

**Reply by kolserdav** [reply]
Надо подумать.
Тут ведь такое дело, типы базы отличаются от типов http...
Вообще идея хорошая, написать им чтобы подумали об авто сериализации специфичных типов. Может что-то подскажут;)

**Reply by atherdon** [reply]
Я думаю что это очень common problem и они могут ее реализовать. тем более что они сразу(призма) позиционировали себя как saas решение. я так понимаю что они хотят пойти по пути Vercel\nextjs - сделать безплатный фреимворк а потом просто продавать хостинг, как heroku

**Comment by kolserdav** [comment]
@vadim9999 по отлову ошибок, мне удалось стандартизировать, в доках опишу, но здесь предупреждаю,
 тип ошибок
 ```json
"errors": [
    {
        "code": number - http статус если 502 то это стандартная `graphql` ошибка и message будут 'Some went wrong',
        "message": string  - в зависимости от статуса либо 'Sww' или персонализированная,
         "status": "error" | "warning" (например если неверный пароль и т.д.)
         "stdErrMessage": "string"
    }
]
```

_Эти изменения вступят в силу в версии 3.2.0_
